File: src/examples/3/example_ws/src/temperature_sensor/src/sensor_node.cpp
/* ----------------------------------------------------------------------------
 * Copyright 2022, Jeferson Lima
 * All Rights Reserved
 * See LICENSE for the license information
 * -------------------------------------------------------------------------- */

/**
 *  @file   examples/3/sensor_node.cpp
 *  @author Jeferson Lima
 *  @brief  Temperature Sensor Example 
 *  @date   Mar 28, 2022
 **/

#include <cstdlib>
#include "ros/ros.h"
#include "std_msgs/Float64.h"


/* @brief minimal ros publish source */
int main(int argc, char **argv)
{

  std_msgs::Float64 sensor;

  ros::init(argc, argv, "temperature_sensor_node");

  ros::NodeHandle n;

  ros::Publisher sensor_pub = n.advertise<std_msgs::Float64>("sensor_node", 10);

  ros::Rate loop_rate(2);

  while(ros::ok())
  {
    // simutale sensor
    sensor.data = 10.0/(rand() % 10 + 1) + 20.0; // 20 -> 30 degree
    ROS_INFO("Temperature: %.2f", sensor.data);

    sensor_pub.publish(sensor);

    ros::spinOnce();

    loop_rate.sleep();
  }

  return 0;
}

File: src/examples/3/example_ws/src/temperature_sensor/script/sensor_node.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import Float64
from numpy import random

def sensor():

    rospy.init_node('sensor_node', anonymous=False)
    pub = rospy.Publisher('sensor/value', Float64, queue_size = 10)
    rate = rospy.Rate(2) #Hz

    value = Float64()

    while not rospy.is_shutdown():

        value.data = 20 + 10*random.rand() # 20 <-> 30 degree
        rospy.loginfo(f"The value is: {value.data}")
        pub.publish(value)

        rate.sleep()


if __name__ == "__main__":
    try:
        sensor()
    except rospy.ROSInterruptException:
        pass
